### ISSUE TITLE
KaTeX not accessible to mobile screen reader users via touch exploration due to the MathML being visually hidden. Visible equation has aria-hidden=true which disables touch exploration. 

### ISSUE THREAD

**Issue by pauljadam** [Khan/KaTeX#820]
The KaTeX demo page has some math equations that render visually hidden MathML output and a visible equation that has aria-hidden=true. What this problem does is prevents mobile screen reader users like VoiceOver for iOS from using touch exploration (explore-by-touch) with their screen reader enabled to move their fingers on the screen and set focus onto the MathML equation. They can't touch the visible equations (non-MathML) directly because the visible equations have aria-hidden=true which prevents screen reader focus.

The screen reader user on iOS can get to the MathML equations but they can only do it through linear navigation, e.g. swipe to the next element one at a time, or use VoiceOver+arrow key navigation. Touch exploration is disabled though and the MathML is skipped over if just moving your finger along the screen trying to focus on the equation. 

My suggestion to solve this problem would be to do something like detecting browsers that support MathML like Safari and then render the MathML as visible equations rather than visually hidden. 

Steps to reproduce:
1. Visit https://khan.github.io/KaTeX/ on an iPhone or iPad
2. Turn on the VoiceOver Screen reader.
3. Set focus with your finger to the Math equations. 

Expected Results:
VoiceOver speaks the Math equations out in an accessible format after you touch them to set focus. 

Actual Results:
VoiceOver makes a "ding, ding, ding" noise as you try to touch the Math equations and you can't actually set focus to the match equations by directly touching them. 

Code that is used to visually hide the accessible MathML:
.katex .katex-mathml {
    position: absolute;
    clip: rect(1px, 1px, 1px, 1px);
    padding: 0;
    border: 0;
    height: 1px;
    width: 1px;
    overflow: hidden
}

Happy to help answer any follow up questions and provide more info!

Thanks!

**Comment by kevinbarabash** [comment]
@pauljadam thanks for reporting this.  Ideally, it be great if we could find a way to make VoiceOver work while showing the HTML rendering.

**Comment by pauljadam** [comment]
@kevinbarabash You could detect iOS or detect Safari and then render the MathML version.

**Comment by edemaine** [comment]
@pauljadam KaTeX generally does server-side rendering, so it can't do any browser detection.  (The exception to this is `auto-render`, which could do this -- but many users don't use `auto-render`.)  Also, this would affect all iOS users, not just those using screen readers, so sighted users would get a much worse rendering experience.

A couple of questions:
* What if we just removed `aria-hidden=true`?  Would that be enough?
* Is this an issue with screen readers in general, or just those on iOS?

**Comment by pauljadam** [comment]
@edemaine the aria-hidden currently is placed on the inaccessible visual math rendering so that would not help the screen reader because they don't want to see the visual rendering, the MathML is not hidden with ARIA it's hidden with CSS visually only. VoiceOver iOS want's to see the MathML not the other rendering that does not have the math semantics. 

The issue is with any mobile screen readers that use explore by touch like VoiceOver for iOS and TalkBack for Android. Chrome is not rendering MathML though so it's not so much an issue on Android as MathML is not supported yet but eventually it will be and would have the same issues as iOS VoiceOver currently. 

Another possibility is some sort of button to render the MathML that is accessible placed right next to the inaccessible visual math rendering?

**Comment by davidflanagan** [comment]
Is there any kind of CSS media query that we could use to determine whether a screen reader is in use? Seem like that would be the ideal way to deal with this.

Failing that, maybe we could solve this (specifically for the Khan webapp) by having a per-user setting that we could query. For KaTeX that would mean we'd want an option to render() to specify visual rendering or a VoiceOver compatible rendering.

**Comment by kevinbarabash** [comment]
@somewhatabstract's suggestion sounds promising and easy to do.  Let's try that first.

**Comment by pauljadam** [comment]
aria- ID references will render the MathML equation as flat static text spoken with no semantics so that won't work. You also can't detect screen readers in web pages.

**Comment by pauljadam** [comment]
http://www.freedomscientific.com/content/html/jawshq/MathML-Samples.html These examples seem to work with explore by touch on iOS, it's using MathJax but I guess the visually hidden MathML content is taking up the same space as the MathJax rendered equation so when you touch it with VoiceOver on the iPad it reads the MathML

**Comment by kevinbarabash** [comment]
@pauljadam the example you linked to is just rendering MathML and nothing else.

**Comment by pauljadam** [comment]
@kevinbarabash  but in Chrome it renders the MathJax. Here's another example that renders MathJax in Chrome but MathML in Safari iOS http://wet-boew.github.io/wet-boew/demos/mathml/mathml-en.html

**Comment by kevinbarabash** [comment]
I think that maybe we can provide some way to optionally render the MathML visually.

**Comment by SmashManiac** [comment]
Here is my perspective as a web developer.

The real issue is that 2 of the major modern browsers, Chrome and Edge, do not support MathML even though they are part of the HTML 5 standard. Note that in both cases, MathML is among their top requested features. If they did implement it, then KaTeX would only need to render a single visible MathML version. This, in my opinion, would be the proper fix for this issue.

Of course, we cannot control Microsoft's priorities for Edge beyond feature requests, but their market share is relatively low and they usually follow behind competitors. For Chrome and its huge market share however, as it is based on the open-source Chromium project, it's only a matter of developers investing the necessary resources in solving that issue. Here is the open Chromium issue for reference: https://bugs.chromium.org/p/chromium/issues/detail?id=6606

I'm not sure what the technical challenges for solving the Chromium issue currently are, or why it's taking so long to do so despite the huge demand for it, but in my opinion this investment should be the number one priority in order to stop wasting countless hours of development and maintenance in alternate methods of rendering math formulas on the web. If people from Khan Academy or other organizations could work with the Chromium developers to resolve this situation, it would be ideal for everyone and a worthy investment for science and education in general.

Regardless of how long it would take for complete MathML browser support, the only plan that seems appropriate to me for KaTeX's development specifically would be to do the following:

1. Update KaTeX's MathML version so it can be visually rendered. (Same as @kevinbarabash suggested, as it would need to be done anyway.)
2. Allow web developers to easily toggle between both KaTeX rendering versions at run time through CSS or JavaScript. (This would enable workarounds similar to what @pauljadam showed as examples earlier.)
3. (Optional) Provide an option for web developers to disable KaTeX's basic HTML rendering at build time. (This would ensure future compatibility and cover use cases where developers only want MathML.)
3. Contribute to MathML's implementation in Chromium. (See above.)
4. Deprecate/remove KaTeX's basic HTML rendering completely. (It would be technical debt at that point.)

You may want to create separate feature issues to tackle some of the stories of that high-level plan.

I hope you found my insights useful.

**Comment by edemaine** [comment]
I don't know MathML super well, but my impression is that it can never render LaTeX math as well as LaTeX does. So I think there will always be an application of rendering LaTeX to a visual language like HTML (or SVG).

**Comment by kevinbarabash** [comment]
The problem with MathML is that each browser is going to render it differently.  It really feels like a stop gap measure.  Although [Houdini](https://ishoudinireadyyet.com) and the [Accessibility Object Model](https://github.com/WICG/aom) are a ways off, they are imo much more promising in terms of moving math rendering and math accessibility forward.  At the bottom of the Chromium issue they're talking about using Custom Layout (part of Houdini) to implement MathML.  When all browsers support Custom Layout, MathML could potentially be provided as a library.

**Comment by ylemkimon** [comment]
IMHO, (La)TeX [is _de facto_ standard for writing and rendering math equations](https://tex.stackexchange.com/questions/49208/why-is-latex-used-as-the-defacto-standard-for-math-equations), at least in science and mathematics. Compare
```latex
x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}
```
with
```xml
<math xmlns="http://www.w3.org/1998/Math/MathML">
 <mi>x</mi>
 <mo>=</mo>
 <mfrac>
  <mrow>
   <mrow>
    <mo>-</mo>
    <mi>b</mi>
   </mrow>
   <mo>&PlusMinus;</mo>
   <msqrt>
    <msup>
     <mi>b</mi>
     <mn>2</mn>
    </msup>
    <mo>-</mo>
    <mrow>
     <mn>4</mn>
     <mo>&InvisibleTimes;</mo>
     <mi>a</mi>
     <mo>&InvisibleTimes;</mo>
     <mi>c</mi>
    </mrow>
   </msqrt>
  </mrow>
  <mrow>
   <mn>2</mn>
   <mo>&InvisibleTimes;</mo>
   <mi>a</mi>
  </mrow>
 </mfrac>
</math>
```

**Comment by SmashManiac** [comment]
Not sure why LaTeX was brought up when KaTeX's documentation only mentions dealing with TeX expressions. In any case, HTML is not a visual language, SVG is not well-suited for math accessibility, and it doesn't matter if a browser renders a math expression slightly differently from another as long as the meaning is the same.

Right now there is only one web standard for mathematics, and it's MathML. Yes, it's verbose and hard to read, but it solves a lot of other issues, and it's the best that we have right now for web output. TeX is great for inputting math formulas though, but it wasn't designed for web output, and that's why projects like KaTeX are valuable for conversion as of now.

Until someone can come off with a better standard than MathML for mathematics on the web, I'm afraid my proposed solution is the best we can do to solve this issue.

**Comment by ylemkimon** [comment]
Sorry, I got confused with the input and the output. But as @edemaine and @kevinbarabash said, its syntax is very limited and styles or polyfills cannot be (easily) controlled.

---

Regarding the original issue, we can investigate [Speech Rule Engine (SRE)](https://github.com/zorkow/speech-rule-engine), which can translate MathML to speech strings, and MathJax uses it in the [accessibility extension](https://github.com/mathjax/MathJax-a11y/). (Related: #457) I've tried it and it successfully converts KaTeX MathML output. The problem is, AFAIK, there is no way to provide an alternative text/description of elements with contents, such as `<div>` and `<span>`, for screenreaders.

**Comment by kevinbarabash** [comment]
I was curious how screen readers actually read out our MathML and it's not pretty.  It reads things as `f`, `left paren`, `x`, `right paren`, etc.  Ideally we'd want it to read `f of x equals the integral...`.

<img width="640" alt="screen shot 2018-09-10 at 11 09 39 pm" src="https://user-images.githubusercontent.com/1044413/45335855-302b7b80-b54f-11e8-899a-ac5e728c22c7.png">

I had to drill down several layers to get to the first real sub-element which it turns our was just `f`.
<img width="596" alt="screen shot 2018-09-10 at 11 13 10 pm" src="https://user-images.githubusercontent.com/1044413/45335849-26097d00-b54f-11e8-82aa-9b8efea75359.png">

Although we have a `<semantics>` tag we only include presentation markup within it.  We should be including content markup which describes the structure of the mathematical expression.  Unfortunately, LaTeX/KaTeX is purely presentation and extracting the structure of a mathematical expression for all possible input strings is impossible.  While it would work for some simple cases, e.g. `\frac{x}{y}` for more complex things it may fail.  There's also the question of whether one tries to build a string from the input or from the macro expanded input.  That being said, doing something is better than doing nothing.  Using a best-effort approach is probably the way to go with this.

In the short term we should probably just set `aria-label` on root to be a string generated by SRE or similar library.  I believe we had some code at one time that generated a string on basic heuristics, I'll see if I can dig it up.

In the long term we should look at generating a semantic tree structure that generally follows what MathML's content markup does.  Then we can use that structure to generate content markup for those browsers that support it generate some other structure such as nested spans without appropriate `role`s and `aria-label`s or maybe the accessibility object model will be ready in some browsers.  The key thing is, is that if we don't a good way of structure data what's read out won't be easily understood let alone easily navigable.

**Comment by kevinbarabash** [comment]
Here's what Khan Academy was using before switching to MathML:
- code: https://github.com/Khan/react-components/blob/master/js/katex-a11y.js
- test cases: https://github.com/Khan/react-components/blob/master/test/katex-a11y-tex.json

Give what we output for MathML I think that what katex-a11y.js generates is superior and we should probably switch back.

**Comment by kevinbarabash** [comment]
I think some of the language could be improved:
- instead of using "start/end superscript", "start/end exponent" might be more appropriate
- `f'` should probably be read as `f prime` instead of `f, prime`
- there are probably a couple of things that could be simplified, e.g. `x^2` could be read as `x squared` instead of `x start superscript 2 end superscript`, same thing for cube root
- for simple fractions like `\frac{5}{12}` it would be nice if we could generate `five twelfths`

It looks like it handles multi-digit numbers correct which is good.

**Comment by kevinbarabash** [comment]
I'm going to be working on porting katex-a11y.js (linked to above) to our contrib folder sometime next week.  We'd like to upgrade khanacademy.org to the latest version of KaTeX but it breaks katex-a11y.js.  As part of that work I'll also be adding tests for it.

**Comment by kevinbarabash** [comment]
I have a WIP diff up for this in #2062.

**Comment by kevinbarabash** [comment]
I'm not sure if this will solve the VoiceOver issues.  I'll do some testing once I figure out how to added this the test page.

**Comment by kevinbarabash** [comment]
I moving this to the backlog. I think we should work to improve accessibility of KaTeX, but this is probably going to take a lot of work and I don't want it blocking a 1.0 release.

**Comment by mladenpanic** [comment]
I am trying to get SR to read katex equations but without success. I don't understand if this feature is currently supported or still work in progress?
I have a few examples of mine, but I've also tried katex.org and I cannot get the screen reader to read any equations. I've been using Chrome and VoiceOver.

I see that MathML is rendered in DOM along with HTML output, but SR is simply ignoring katex altogether.

Any ideas?

**Comment by ylemkimon** [comment]
@mladenpanic Currently, directly selecting the KaTeX element doesn't work. However, scrolling through elements should work.

**Comment by mladenpanic** [comment]
I have tried reading all content on the page but VoiceOver skips katex parts. 
Also, I am using katex equations as labels for inputs, multiple-choice questions using radio buttons, SR never reads the rendered equation within <label> again, it just skips it.

I am required to meet strict accessibility standards so I am wondering is there any way to make this work, or should I try mathjax. I prefer katex because of its speed, but if I cannot get it to be recognized by SR I will need to return to mathjax.

**Comment by ylemkimon** [comment]
@mladenpanic I don't have an iPhone to test it now, but with Google TalkBack, MathML reading seems to be broken somehow. I'll investigate more.

Meanwhile, there is a plugin https://github.com/KaTeX/KaTeX/tree/master/contrib/render-a11y-string, which can build an alternative text.

**Comment by necrohell** [comment]
Adding to this -- NVDA reader also cannot see the rendered math on the KaTeX homepage.

**Comment by UlyssesZh** [comment]
I suggest:

1. Move `renderA11yString` from the contrib to the base library so that functions in the base library can use it. It only increases the script size by 3%.
2. Add a boolean config option `aria`. When it is true, the output will contain more accessibility-related attributes. Then VoiceOver can work for `html` and `mathml` output, but still not `htmlAndMathml` output.
   - For `html` output, there will be no `aria-hidden`, and `aria-label` will be set using `renderA11yString`, and also `role="math"`.
   - For `mathml` output, `aria-label` will be set using `renderA11yString`.
   - For `htmlAndMathml` output, `aria-label` will be added to the MathML element, and the HTML element has `aria-hidden` and `role="math"`. Note that at least one of them has to have `aria-hidden`, otherwise accessibility is negatively impacted by seeing both elements.

This is now something easy to do given that we already have `renderA11yString` implemented. @k4b7 If you think this is okay, I can work on a PR.

**Comment by grigoriy-reshetniak** [comment]
I've checked current state of this. To be precise, OP was talking about **mobile** accessibility, not VoiceOver in general. @mladenpanic, your issue most likely coming from using Tab navigation. When I use VO + -> (VoiceOver modificator, `Caps Lock` by default, and left arrow) it reads hidden MathML just fine while using Chrome and VoiceOver.

Test environment:
- https://katex.org/
- Output: `htmlAndMathml`.
- VoiceOver while using Safari for iOS 26.4.1 
- TalkBack 15.1.02.0, Chrome for Android 147.0.7727.50

#### VoiceOver
Equation doesn't produce any speech output if tapped directly, but if items on page read one by one -- VO picks up hidden MathML and output is identical with desktop one.

#### TalkBack
Equation doesn't produce any speech output if tapped directly, but if items on page read one by one -- TB picks it symbol by symbol and ignores MathML completely.


It makes sense to embed `renderA11yString`, however I'm not sure if the `htmlAndMathml` output mode would even make sense when `aria-label` is present. @edemaine what is your opinion on this?

**Comment by UlyssesZh** [comment]
The problem is about pointer-based screen readers in general. If you do not point at the element that you want it to read, it is out of the scope of OP's issue.

> It makes sense to embed `renderA11yString`, however I'm not sure if the `htmlAndMathml` output mode would even make sense when `aria-label` is present.

Yeah it is useless from the point of view of screen readers. I imagine the MathML is still useful for the `copy-tex` plugin, but we can put the TeX code in a `data-tex` field of the HTML element to make it still work.

c.f. #593

**Comment by UlyssesZh** [comment]
I saw this discussion from @alexander-turner (https://github.com/KaTeX/KaTeX/pull/4162#issuecomment-4140116453):

> As for html-only, I'm now thinking—`aria-label` is inferior to the MathML anyways, so why expand the bundle size by supporting it? I'm thinking we can just log a warning that if the HTML will be consumed by screen readers, the caller should use the "htmlAndMathML" setting. WDYT?

We may reconsider this in the context of the current issue. Should we agree that hidden MathML is inferior to HTML-only with `aria-label` for VoiceOver?
